### ISSUE TITLE
Status of schedBundle.sc?

### ISSUE THREAD

**Issue by jamshark70** [supercollider/supercollider#4834]
schedBundle.sc seems to have been transferred over from the crucial library some years ago. It seems to be part of the OSCBundle/MixedBundle protocol[1] -- JITLib uses both of these classes (but I'm pretty sure they're not widely used elsewhere), and JITLib contains no references to `schedBundle`.

There are some issues:

- Use of `Main.elapsedTime` for scheduling bundles. We should always timestamp bundles based on logical time, never physical time!

- Integer:schedBundle includes `latencyBeats = Tempo.secs2beats(server.latency ? 0.05);`. `Tempo` is defined in the crucial library quark. So we have a main library method that silently requires an external dependency.

`schedBundle` is used in the main library only once (OSCBundle:sendAtTime), and sendAtTime is not used anywhere in the main library (though it's used extensively in the crucial library).

Over a decade later, I'm not sure why this was moved into the main library, but maybe it's time to consider moving it back? It was probably intended to support OSCBundle better, in the hope of OSCBundle/MixedBundle becoming widely used, but that hasn't happened.

[1] Some `schedBundle` implementations call `bundle.doPrepare`, which in the main class library is defined only on OSCBundle. So I'm confident in saying that these methods are not about bundle arrays in general.

**Comment by telephon** [comment]
Yes, `schedBundle` may have been there to provide a standard protocol. Do you see a possibility to fix it instead of removing it? Or, well, that in turn may break crucial lib that depends on implementation details.

I also believ nobody uses it because it wouldn't work without crucial lib.

**Comment by jamshark70** [comment]
Let me also bring @crucialfelix into the conversation.

Chris -- I was wondering what to do with schedBundle.sc -- summarizing quickly from above, IIRC it was moved into the main class library from crucial library in the hopes of MixedBundle becoming widely used. But that didn't happen, and in particular `Integer:schedBundle` (main library) depends on `Tempo` from crucial.

I guess the possible ways forward are:

- Move schedBundle.sc back into the crucial library. (Possible user pain when upgrading -- if they use Patch a lot, and upgrade SC without upgrading crucial, then `schedBundle` is likely to fail.)

- Fix schedBundle.sc to remove external dependencies (and also schedule by logical time rather than elapsed beats).

- Or, add a comment to the file that this is legacy code that hasn't been maintained (use at own risk yadda yadda).

I'm kind of leaning towards moving it back to crucial. Fixing it would change behavior, introducing risk to the crucial library, and also, there's the benefit of cleaning up usages that never took off. But I don't have a strong objection to fixing it either.

(The thing that brought up this question was a forum thread, where someone thought that Event bundles are scheduled based on elapsedBeats because schedBundle.sc refers to elapsedBeats, and the string "schedBundle" appears throughout Event.sc -- when in fact "schedBundle" in Event.sc is an entirely different function. If it's introducing confusion in understanding class library behavior, and it's not in wide usage, then it may have outlived its usefulness.)

**Comment by crucialfelix** [comment]
Those methods are—as stated in the header comment—support methods for `atTime` arguments

```
1.schedBundle
0.25.schedBundle
nil.schedBundle
date.schedBundle
```

I don't remember how or why there were moved into common. I certainly didn't do it. History was erased (JMC tried to warn us lol).

It's back in crucial library now: https://github.com/crucialfelix/crucial-library/blob/develop/Scheduling/schedBundle.sc

Remind me - is there a compile error if both crucial and the common library include this identical file of extensions?

**Comment by snappizz** [comment]
maybe this history will help investigation? https://github.com/supercollider/supercollider/commits/66266a2296d6931a50e002d4ea23271525536fdc/build/SCClassLibrary/Common/Control/schedBundle.sc

**Comment by crucialfelix** [comment]
Ah, so it was me. 

I think the main thing we wanted to move was MixedBundle.

there is still `OSCBundle.sendAtTime` which uses those methods.

I'm not using crucial-library or supercollider myself anymore (use supercollider.js when I have time to play), but I can easily update it so it would not depend on any common lib changes.

Over all though, is any of this really important?

**Comment by jamshark70** [comment]
> Remind me - is there a compile error if both crucial and the common library include this identical file of extensions?

Shouldn't be a compile error. There might be a warning somewhere but I think it's not fatal. (I have crucial library installed; if it were fatal, I think I would have seen it.)

> Overall though, is any of this really important?

Probably not, but it's a move toward cleaning up.

If it were only the elapsedBeats thing, I guess I'd leave it. Undocumented dependencies in the main class library should be tidied up though (IMO the main class library should be self-sufficient -- extensions may require other extensions, but the main class library should stand on its own). I guess that's easily fixed by replacing Tempo with `TempoClock.tempo` or by checking to see if `thisThread.clock` answers tempo and using that if possible.

**Comment by jamshark70** [comment]
In any case, it would be rude to tamper with your code without any heads-up at all. (That's the main reason to @ you on this.)
